tensorboard/backend/path_prefix_test.py

```diff
@@ -0,0 +1,112 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Tests for `tensorboard.backend.path_prefix`."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import json
+
+import werkzeug
+
+from tensorboard import errors
+from tensorboard import test as tb_test
+from tensorboard.backend import path_prefix
+
+
+class PathPrefixMiddlewareTest(tb_test.TestCase):
+  """Tests for `PathPrefixMiddleware`."""
+
+  def _echo_app(self, environ, start_response):
+    # https://www.python.org/dev/peps/pep-0333/#environ-variables
+    data = {
+        "path": environ.get("PATH_INFO", ""),
+        "script": environ.get("SCRIPT_NAME", ""),
+    }
+    body = json.dumps(data, sort_keys=True)
+    start_response("200 OK", [("Content-Type", "application/json")])
+    return [body]
+
+  def _assert_ok(self, response, path, script):
+    self.assertEqual(response.status_code, 200)
+    actual = json.loads(response.get_data())
+    expected = dict(path=path, script=script)
+    self.assertEqual(actual, expected)
+
+  def test_bad_path_prefix_without_leading_slash(self):
+    with self.assertRaises(ValueError) as cm:
+      path_prefix.PathPrefixMiddleware(self._echo_app, "hmm")
+    msg = str(cm.exception)
+    self.assertIn("must start with slash", msg)
+    self.assertIn(repr("hmm"), msg)
+
+  def test_bad_path_prefix_with_trailing_slash(self):
+    with self.assertRaises(ValueError) as cm:
+      path_prefix.PathPrefixMiddleware(self._echo_app, "/hmm/")
+    msg = str(cm.exception)
+    self.assertIn("must not end with slash", msg)
+    self.assertIn(repr("/hmm/"), msg)
+
+  def test_empty_path_prefix(self):
+    app = path_prefix.PathPrefixMiddleware(self._echo_app, "")
+    server = werkzeug.test.Client(app, werkzeug.BaseResponse)
+
+    with self.subTest("at empty"):
+      self._assert_ok(server.get(""), path="", script="")
+
+    with self.subTest("at root"):
+      self._assert_ok(server.get("/"), path="/", script="")
+
+    with self.subTest("at subpath"):
+      response = server.get("/foo/bar")
+      self._assert_ok(server.get("/foo/bar"), path="/foo/bar", script="")
+
+  def test_nonempty_path_prefix(self):
+    app = path_prefix.PathPrefixMiddleware(self._echo_app, "/pfx")
+    server = werkzeug.test.Client(app, werkzeug.BaseResponse)
+
+    with self.subTest("at root"):
+      response = server.get("/pfx")
+      self._assert_ok(response, path="", script="/pfx")
```

Comment:
no action required; I simply dunno: how does our server respond when the path is "" vs. "/"?

Reply:
Good question. The answer is that Werkzeug normalizes that out, so our
app (which uses Werkzeug for routing) only sees `request.path == "/"`:

```python
# main.py
import werkzeug

def bare_app(environ, start_response):
  start_response("200 OK", [("Content-Type", "text/plain")])
  return [repr(environ.get("PATH_INFO")).encode("utf-8")]

@werkzeug.Request.application
def werkzeug_app(request):
  return werkzeug.Response(repr(request.path), content_type="text/plain")

start_response = lambda status, headers: None
request_empty = {"REQUEST_METHOD": "GET", "PATH_INFO": ""}
request_root = {"REQUEST_METHOD": "GET", "PATH_INFO": "/"}

print("With base WSGI:")
print(list(bare_app(request_empty, start_response)))
print(list(bare_app(request_root, start_response)))

print("With Werkzeug:")
print(list(werkzeug_app(request_empty, start_response)))
print(list(werkzeug_app(request_root, start_response)))
```

```
$ python main.py
With base WSGI:
[b"''"]
[b"'/'"]
With Werkzeug:
[b"'/'"]
[b"'/'"]
```